File: apps/mobile/modules/siyue-offline-clock/android/src/main/java/app/siyue/offlineclock/SiyueOfflineClockModule.kt
package app.siyue.offlineclock

import android.os.SystemClock
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.util.UUID

class SiyueOfflineClockModule : Module() {
  companion object {
    private val processEpoch: String = UUID.randomUUID().toString().lowercase()
  }

  override fun definition() = ModuleDefinition {
    Name("SiyueOfflineClock")

    Function("snapshot") {
      mapOf(
        "bootId" to processEpoch,
        "elapsedRealtimeMs" to SystemClock.elapsedRealtime().toDouble()
      )
    }
  }
}

File: apps/mobile/modules/siyue-offline-clock/ios/SiyueOfflineClockModule.swift
import Darwin
import ExpoModulesCore

private let processEpoch = UUID().uuidString.lowercased()

private func continuousMilliseconds() -> Double {
  var timebase = mach_timebase_info_data_t()
  mach_timebase_info(&timebase)
  let nanoseconds = Double(mach_continuous_time()) * Double(timebase.numer) / Double(timebase.denom)
  return nanoseconds / 1_000_000
}

public class SiyueOfflineClockModule: Module {
  public func definition() -> ModuleDefinition {
    Name("SiyueOfflineClock")

    Function("snapshot") {
      return [
        "bootId": processEpoch,
        "elapsedRealtimeMs": continuousMilliseconds()
      ]
    }
  }
}
